packages/components/src/sites-table/use-sites-table-filtering.tsx

```diff
@@ -1,77 +1,75 @@
 import { useFuzzySearch } from '@automattic/search';
 import { useI18n } from '@wordpress/react-i18n';
 import { useMemo } from 'react';
-// eslint-disable-next-line no-restricted-imports
-import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+import { SiteObjectWithStatus, getSiteStatus, siteStatuses } from './site-status';
 
-export type SiteStatus = 'all' | 'coming-soon' | 'private' | 'public';
+export const DEFAULT_SITE_STATUS_FILTER_VALUE = 'all';
+
+export const siteStatusFilterValues = [
+	DEFAULT_SITE_STATUS_FILTER_VALUE,
+	...siteStatuses,
+] as const;
+
+export type FilterableSiteStatuses = typeof siteStatusFilterValues[ number ];
 
 interface SitesTableFilterOptions {
-	status?: string;
+	status: FilterableSiteStatuses;
 	search?: string;
 }
 
 interface Status {
 	title: React.ReactChild;
-	name: SiteStatus;
+	name: FilterableSiteStatuses;
 	count: number;
 }
 
-interface UseSitesTableFilteringResult {
-	filteredSites: SiteExcerptData[];
+interface UseSitesTableFilteringResult< T > {
+	filteredSites: T[];
 	statuses: Status[];
 }
 
-export function useSitesTableFiltering(
-	allSites: SiteExcerptData[],
-	{ status = 'all', search }: SitesTableFilterOptions
-): UseSitesTableFilteringResult {
+type SiteObjectWithBasicInfo = SiteObjectWithStatus & {
+	URL: string;
+	name: string;
+	slug: string;
+};
+
+export function useSitesTableFiltering< T extends SiteObjectWithBasicInfo >(
+	allSites: T[],
+	{ status, search }: SitesTableFilterOptions
```

Comment:
Previously, we were accepting any arbitrary string (or even nothing at all) as the value for `status`, and that's problematic because it loses constraints so it was harder to be exhaustive when grouping.

It's easy to visualize that and how the type system is sounder inside this function now: try adding a new entry in `siteStatusFilterValues` and check how the grouping and filtering fail. Unfortunately, TS doesn't have pattern matching included so it's not possible to perform exhaustive matching in `const statuses`, but that's good enough for now.